pkg/providers/v1/instances_v2.go

```diff
@@ -0,0 +1,81 @@
+// This file implements the InstancesV2 interface.
+// InstancesV2 is an abstract, pluggable interface for cloud provider instances.
+// Unlike the Instances interface, it is designed for external cloud providers and should only be used by them.
+
+package aws
+
+import (
+	"context"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	cloudprovider "k8s.io/cloud-provider"
+)
+
+// InstancesV2.InstanceExists
+// InstanceExists returns true if the instance for the given node exists according to the cloud provider.
+// Use the node.name or node.spec.providerID field to find the node in the cloud provider.
+func (c *Cloud) InstanceExists(ctx context.Context, node *v1.Node) (bool, error) {
+	providerID, err := c.getProviderID(ctx, node)
+	if err != nil {
+		return false, err
+	}
+
+	return c.InstanceExistsByProviderID(ctx, providerID)
+}
+
+// InstancesV2.InstanceShutdown
+// InstanceShutdown returns true if the instance is shutdown according to the cloud provider.
+// Use the node.name or node.spec.providerID field to find the node in the cloud provider.
+func (c *Cloud) InstanceShutdown(ctx context.Context, node *v1.Node) (bool, error) {
+	providerID, err := c.getProviderID(ctx, node)
+	if err != nil {
+		return false, err
+	}
+
+	return c.InstanceShutdownByProviderID(ctx, providerID)
+}
+
+// InstancesV2.InstanceMetadata
+// InstanceMetadata returns the instance's metadata. The values returned in InstanceMetadata are
+// translated into specific fields and labels in the Node object on registration.
+// Implementations should always check node.spec.providerID first when trying to discover the instance
+// for a given node. In cases where node.spec.providerID is empty, implementations can use other
+// properties of the node like its name, labels and annotations.
+func (c *Cloud) InstanceMetadata(ctx context.Context, node *v1.Node) (*cloudprovider.InstanceMetadata, error) {
+	providerID, err := c.getProviderID(ctx, node)
+	if err != nil {
+		return &cloudprovider.InstanceMetadata{}, err
+	}
+
+	instanceType, err := c.InstanceTypeByProviderID(ctx, providerID)
+	if err != nil {
+		return &cloudprovider.InstanceMetadata{}, err
+	}
+
+	zone, err := c.GetZoneByProviderID(ctx, providerID)
+	if err != nil {
+		return &cloudprovider.InstanceMetadata{}, err
+	}
+
+	nodeAddresses, err := c.NodeAddressesByProviderID(ctx, providerID)
+	if err != nil {
+		return &cloudprovider.InstanceMetadata{}, err
+	}
+
+	return &cloudprovider.InstanceMetadata{
+		ProviderID:    providerID,
+		InstanceType:  instanceType,
+		NodeAddresses: nodeAddresses,
+		Zone:          zone.FailureDomain,
+		Region:        zone.Region,
+	}, nil
+}
+
+func (c *Cloud) getProviderID(ctx context.Context, node *v1.Node) (string, error) {
+	instanceID, err := c.InstanceID(ctx, types.NodeName(node.Name))
+	if err != nil {
+		return "", err
+	}
+	return c.ProviderName() + "://" + instanceID, nil
```

Comment:
There should already be an impl for this in `aws.go`, and the provider ID should be suffixed by the AZ

Reply:
One would think that... But it's not.

> the provider ID should be suffixed by the AZ

Ya, I knew I needed to update this, but ran out of time. I just copied the code from k8s/k8s to get the provider ID, so I was a little confused by it. I'll dig in further and make sure it returns the right output with some unit tests.

Reply:
Ok, I was a little confused by because I copied the logic upstream ([ref 1]() [ref 2]()) and didn't totally close the loop on it. However, looks like this is actually the right behavior. In totally obvious fashion, `instanceID`, in this case, is actually NOT the instance ID. It's the instance ID plus everything else we tack on the provider ID...

https://github.com/kubernetes/cloud-provider-aws/blob/2fc845dd863907dd5f74d10b8de9575a7418af27/pkg/providers/v1/aws.go#L1773-L1788

Reply:
Am I missing something, or does this function ignore the provider ID already set on the node resource?
At least some installers set the provider ID through the kubelet flag, which in turn sets it on the node resource as it is created.

Reply:
Hm looks like the provider ID returned here won't overwrite an existing one: https://github.com/kubernetes/kubernetes/blob/5031ae11590748f247c15c81f9f35390a786104b/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L524-L530

But we could save some EC2 calls if we use the existing provider ID.

Reply:
Probably won't save any calls to EC2 because we use cached values and will need the describe values for the other fields, but getting it directly, if it's set, makes sense.